zyte_api/_retry.py

```diff
@@ -177,3 +222,130 @@ def build(self) -> AsyncRetrying:
+        assert retry_state.outcome, "Unexpected empty outcome"
+        exc = retry_state.outcome.exception()
+        assert exc, "Unexpected empty exception"
+        if exc.status == 521:  # type: ignore
+            retry_state.counter["permanent_download_error"] += 1  # type: ignore
+            if retry_state.counter["permanent_download_error"] >= self._max_permanent:  # type: ignore
+                return True
+        retry_state.counter["download_error"] += 1  # type: ignore
+        if retry_state.counter["download_error"] >= self._max_total:  # type: ignore
+            return True
+        return False
+
+
+class stop_on_uninterrupted_status(stop_base):
+    """Stop after the specified max number of error responses with the same
+    status code in a row."""
+
+    def __init__(self, _max: int) -> None:
+        self._max = _max
+
+    def __call__(self, retry_state: "RetryCallState") -> bool:
+        assert retry_state.outcome, "Unexpected empty outcome"
+        exc = retry_state.outcome.exception()
+        assert exc, "Unexpected empty exception"
+        count = 0
+        for status in reversed(retry_state.status_history):  # type: ignore
+            if status == exc.status:  # type: ignore
+                count += 1
+                if count >= self._max:
+                    return True
+            elif status not in {-1, 429, 503}:
+                return False
+        return False
+
+
+class AggressiveRetryFactory(RetryFactory):
+    """Factory class that builds the :class:`tenacity.AsyncRetrying` object
+    that defines the :ref:`aggressive retry policy <aggressive-retry-policy>`.
+
+    To create a custom retry policy, you can subclass this factory class,
+    modify it as needed, and then call :meth:`build` on your subclass to get
+    the corresponding :class:`tenacity.AsyncRetrying` object.
+
+    For example, to double the maximum number of attempts for all error
+    responses and double the time network errors are retried:
+
+    .. code-block:: python
+
+        from zyte_api import (
+            AggressiveRetryFactory,
+            stop_after_uninterrupted_delay,
+            stop_on_download_error,
+            stop_on_uninterrupted_status,
+        )
+
+
+        class CustomRetryFactory(AggressiveRetryFactory):
+            download_error_stop = stop_on_download_error(max_total=16, max_permanent=8)
+            network_error_stop = stop_after_uninterrupted_delay(30 * 60)
+            undocumented_error_stop = stop_on_uninterrupted_status(8)
+
+
+        CUSTOM_RETRY_POLICY = CustomRetryFactory().build()
+    """
+
+    retry_condition = (
+        RetryFactory.retry_condition
+        | retry_if_exception(_download_error)
+        | retry_if_exception(_undocumented_error)
+    )
+
+    download_error_stop = stop_on_download_error(max_total=8, max_permanent=4)
+    download_error_wait = RetryFactory.temporary_download_error_wait
+
+    undocumented_error_stop = stop_on_uninterrupted_status(4)
```

Comment:
Every use of stop_on_uninterrupted_status has a risk: if we're not careful (or if API behavior changes), and there are ABABAB exceptions, where A uses stop_on_uninterrupted_status, we never stop retrying A. 

Is there a strong reason to use uninterrupted here?